dvc/command/completion.py

```diff
@@ -0,0 +1,203 @@
+  compgen -f -- $1
+  compgen -d -S '/' -- $1  # recurse into subdirs
+}
+
+# $1=COMP_WORDS[1]
+_dvc_replace_hyphen() {
+  echo $1 | sed 's/-/_/g'
+}
+
+# $1=COMP_WORDS[1]
+_dvc_compgen_command() {
+  local flags_list="_dvc_$(_dvc_replace_hyphen $1)"
+  local args_gen="${flags_list}_COMPGEN"
+  COMPREPLY=( $(compgen -W "$_dvc_global_options ${!flags_list}" -- "$word"; \
+[ -n "${!args_gen}" ] && ${!args_gen} "$word") )
+}
+
+# $1=COMP_WORDS[1]
+# $2=COMP_WORDS[2]
+_dvc_compgen_subcommand() {
+  local flags_list="_dvc_$(_dvc_replace_hyphen $1)_$(_dvc_replace_hyphen $2)"
+  local args_gen="${flags_list}_COMPGEN"
+  [ -n "${!args_gen}" ] && local opts_more="$(${!args_gen} "$word")"
+  local opts="${!flags_list}"
+  if [ -z "$opts$opts_more" ]; then
+    _dvc_compgen_command $1
+  else
+    COMPREPLY=( $(compgen -W "$_dvc_global_options $opts" -- "$word"; \
+[ -n "$opts_more" ] && echo "$opts_more") )
+  fi
+}
+
+# Notes:
+# `COMPREPLY` contains what will be rendered after completion is triggered
+# `word` refers to the current typed word
+# `${!var}` is to evaluate the content of `var`
+# and expand its content as a variable
+#       hello="world"
+#       x="hello"
+#       ${!x} ->  ${hello} ->  "world"
+_dvc() {
+  local word="${COMP_WORDS[COMP_CWORD]}"
+
+  COMPREPLY=()
+
+  if [ "${COMP_CWORD}" -eq 1 ]; then
+    case "$word" in
+      -*) COMPREPLY=($(compgen -W "$_dvc_options" -- "$word")) ;;
+      *) COMPREPLY=($(compgen -W "$_dvc_commands" -- "$word")) ;;
+    esac
+  elif [ "${COMP_CWORD}" -eq 2 ]; then
+    _dvc_compgen_command ${COMP_WORDS[1]}
+  elif [ "${COMP_CWORD}" -ge 3 ]; then
+    _dvc_compgen_subcommand ${COMP_WORDS[1]} ${COMP_WORDS[2]}
+  fi
+
+  return 0
+}
+
+complete -o nospace -F _dvc dvc""",
+        file=fd,
+    )
+
+
+class CmdCompletion(CmdBase):
+    def run(self):
+        from dvc.cli import get_main_parser
+
+        if self.args.output == "-":
+            logger.debug("Writing bash completion to stdout")
+            fd = sys.stdout
+        else:
+            logger.info(f"Writing bash completion to {self.args.output}")
+            fd = open(self.args.output, "w")
+        parser = get_main_parser()
+        print_bash(parser, fd)
+        del fd
+        return 0
+
+
+def add_parser(subparsers, parent_parser):
+    COMPLETION_HELP = "Enable shell tab completion."
+    COMPLETION_DESCRIPTION = (
+        "Automatically generates shell tab completion script."
+    )
+    completion_parser = subparsers.add_parser(
+        "completion",
+        parents=[parent_parser],
+        description=append_doc_link(COMPLETION_DESCRIPTION, "completion"),
+        help=COMPLETION_HELP,
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+    )
+    completion_parser.add_argument(
+        "-o",
+        "--output",
+        help="Output filename for completion script. Use - for stdout.",
+        metavar="<path>",
+        default="-",
+    )
+    completion_parser.set_defaults(func=CmdCompletion)
```

Comment:
You will have `zsh` soon and probably something else, so you should account for switching between those when you create a new command. Otherwise, it might break backward compatibility soon.

Reply:
added `--shell {bash,dvc}`